engine/interpreter/src/main/scala/pl/touk/nussknacker/engine/dict/ProcessDictSubstitutor.scala

```diff
@@ -0,0 +1,94 @@
+      val nodeExpressionId = exprIdWithMetadata.expressionId
+      val nodeTypingInfo = processTypingInfo.getOrElse(nodeExpressionId.nodeId.id, Map.empty)
+      val optionalExpressionTypingInfo = nodeTypingInfo.get(nodeExpressionId.expressionId)
+      val substitutedExpression = optionalExpressionTypingInfo.flatMap(prepareSubstitutionsCollector(_, replacingStrategy)).map { substitutionsCollector =>
+        val substitutions = substitutionsCollector.collectSubstitutions(expr)
+        val afterSubstitution = ExpressionSubstitutor.substitute(expr.expression, substitutions)
+        if (substitutions.nonEmpty)
+          logger.debug(s"Found ${substitutions.size} substitutions in expression: ${process.metaData.id}>${nodeExpressionId.nodeId.id}>${nodeExpressionId.expressionId}. " +
+            s"Expression: '${expr.expression}' replaced with '$afterSubstitution'")
+        afterSubstitution
+      }.getOrElse(expr.expression)
+      expr.copy(expression = substitutedExpression)
+    }
+
+    rewriter.rewriteProcess(process)
+  }
+
+  def reversed: ProcessDictSubstitutor = new ProcessDictSubstitutor(KeyToLabelReplacingStrategy, prepareSubstitutionsCollector)
+
+}
+
+object ProcessDictSubstitutor extends LazyLogging {
+
+  def apply(): ProcessDictSubstitutor = {
+    new ProcessDictSubstitutor(LabelToKeyReplacingStrategy, prepareSubstitutionsCollector)
+  }
+
+  private def prepareSubstitutionsCollector(typingInfo: ExpressionTypingInfo, replacingStrategy: ReplacingStrategy) = typingInfo match {
+    case SpelExpressionTypingInfo(intermediateResults) =>
+      Some(new SpelSubstitutionsCollector(n => intermediateResults.get(SpelNodeId(n)), replacingStrategy))
+    case _ =>
+      None
+  }
+
+  trait BaseReplacingStrategy extends ReplacingStrategy {
+
+    def findReplacement(typedNodeTree: List[TypedTreeLevel]): Option[String] = typedNodeTree match {
+      case
+        TypedTreeLevel(OptionallyTypedNode(indexerKey: StringLiteral, _) :: Nil) ::
+          TypedTreeLevel(OptionallyTypedNode(_: Indexer, _) :: OptionallyTypedNode(_, Some(static: StaticTypedDict)) :: _) :: _ =>
+        val indexerKeyValue = indexerKey.getLiteralValue.getValue.toString
+        val replacement = findStaticDictReplacement(static, indexerKeyValue)
+
+        replacement.map(key => s"'$key'")
+      case _ => None
+    }
+
+    protected def findStaticDictReplacement(static: StaticTypedDict, indexerKeyValue: String): Option[String]
+
+  }
+
+  object LabelToKeyReplacingStrategy extends BaseReplacingStrategy {
+
+    override protected def findStaticDictReplacement(static: StaticTypedDict, indexerKeyValue: String): Option[String] = {
+      val replacement = static.keysByLabel.get(indexerKeyValue)
+      if (replacement.isEmpty) {
+        logger.warn(s"Can't find key for label: $indexerKeyValue in dict: $static. Probable change in dict definition. Will be used label in this place.")
```

Comment:
Is it quite ok to do this. We probably should add some warning that will be displayed in process, but it shouldn't be blocking error because it could be the situation that some key->label pair was missing for a while. With current implementation it will be fine, because key wouldn't be replaced by label during "reverse resolving" process and then also wouldn't be replaced by key during "resolving" process. At the end we will have the same "key" in saved process and it will be fine after key->label pair come back. 
All this logic will be more clear if we have something like ResolvedExpression instead of Expression in our model and there will be "original" (unresolved) expression and list of substitutions, but I don't want to change model on this phase of incubation of this feature...